File: capturer/cache_diagnostics_test.go
package capturer

import (
	"bytes"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func diagnosticRecorder(t *testing.T) *Recorder {
	t.Helper()
	r, err := New(Config{Mode: "mekugi"})
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = r.Close() })
	return r
}

func TestCacheFingerprintsArePrivateAndFramingIndependent(t *testing.T) {
	r := diagnosticRecorder(t)
	a := r.requestFingerprint([]byte(`{"model":"model","instructions":"private <instructions>","prompt_cache_key":"private key","input":[{"content":"private <text>","n":1000}]}`))
	b := r.requestFingerprint([]byte(` { "input" : [ { "n":1e3, "content":"private \u003ctext\u003e" } ], "prompt_cache_key":"private key", "instructions":"private \u003cinstructions\u003e", "model":"model" } `))
	if got := comparePrefix(a, b); got.Status != "identical" {
		t.Fatalf("framing changed fingerprint: %+v", got)
	}
	if a.RequestKey != b.RequestKey {
		t.Fatal("stable request cache key changed")
	}
	encoded, _ := json.Marshal(a)
	for _, secret := range []string{"private", "<text>", "<instructions>"} {
		if strings.Contains(string(encoded), secret) {
			t.Fatal("raw content retained in fingerprint")
		}
	}
	other := diagnosticRecorder(t).requestFingerprint([]byte(`{"model":"model","input":[]}`))
	if got := comparePrefix(a, other); got.Status != "unavailable" {
		t.Fatal("fingerprints compared across recorder lifetimes")
	}
	clone := cloneFingerprint(a)
	clone.Items[0] = "mutated"
	clone.Fields["model"] = "mutated"
	if a.Items[0] == clone.Items[0] || a.Fields["model"] == clone.Fields["model"] {
		t.Fatal("snapshot aliases recorder fingerprint state")
	}
}

func TestCacheFingerprintLocatesChangesAndBoundsHistory(t *testing.T) {
	r := diagnosticRecorder(t)
	base := r.requestFingerprint([]byte(`{"model":"model","input":["one","two"]}`))
	appended := r.requestFingerprint([]byte(`{"input":["one","two","three"],"model":"model"}`))
	if got := comparePrefix(base, appended); got.Status != "appended" || got.CommonItems != 2 {
		t.Fatalf("append: %+v", got)
	}
	changed := r.requestFingerprint([]byte(`{"model":"different","input":["one","changed"]}`))
	if got := comparePrefix(base, changed); got.Status != "changed" || got.CommonItems != 1 || len(got.ChangedFields) != 1 || got.ChangedFields[0] != "model" {
		t.Fatalf("change: %+v", got)
	}
	shortened := r.requestFingerprint([]byte(`{"model":"model","input":["one"]}`))
	if got := comparePrefix(base, shortened); got.Status != "changed" {
		t.Fatalf("removed history: %+v", got)
	}
	items := make([]string, maxFingerprintItems+1)
	encoded, _ := json.Marshal(map[string]any{"input": items})
	truncated := r.requestFingerprint(encoded)
	if truncated.Complete || len(truncated.Items) != maxFingerprintItems || truncated.ItemCount != len(items) {
		t.Fatal("unbounded fingerprint history")
	}
	if comparePrefix(truncated, truncated).Status != "unavailable" {
		t.Fatal("partial fingerprint claimed a stable prefix")
	}
}

func TestCacheDiagnosisUsesArrivalOrderThreadAndFinalAttempt(t *testing.T) {
	r := diagnosticRecorder(t)
	first := r.requestFingerprint([]byte(`{"model":"model","input":["one"]}`))
	first.RoutingKey = "route-a"
	second := r.requestFingerprint([]byte(`{"model":"model","input":["one","two"]}`))
	second.RoutingKey = "route-b"
	exchanges := []exchangeMetrics{
		{Sequence: 2, PredecessorSequence: 1, ThreadID: "thread", Status: "completed", ClientFingerprint: second, ProviderAttempts: []providerAttemptMetrics{{Fingerprint: first}, {Fingerprint: second, ProjectedFingerprint: second}}},
		{Sequence: 1, ThreadID: "thread", Status: "completed", ClientFingerprint: first, ProviderAttempts: []providerAttemptMetrics{{Fingerprint: first, ProjectedFingerprint: first}}},
		{Sequence: 3, ThreadID: "other", Status: "completed", ClientFingerprint: second, ProviderAttempts: []providerAttemptMetrics{{Fingerprint: second, ProjectedFingerprint: second}}},
	}
	diagnoseCacheExchanges(exchanges)
	got := exchanges[0].CacheDiagnosis
	if got.PreviousSequence != 1 || got.Provider.Status != "appended" || got.Projected.Status != "appended" || got.Routing != "changed" {
		t.Fatalf("diagnosis: %+v", got)
	}
	if exchanges[2].CacheDiagnosis.PreviousSequence != 0 || exchanges[2].CacheDiagnosis.Provider.Status != "unavailable" {
		t.Fatal("unrelated thread became cache predecessor")
	}
}

func TestCacheDiagnosisDoesNotSkipPendingPredecessor(t *testing.T) {
	r := diagnosticRecorder(t)
	fp := r.requestFingerprint([]byte(`{"input":["stable"]}`))
	exchanges := []exchangeMetrics{
		{Sequence: 1, ThreadID: "thread", Status: "completed", ProviderAttempts: []providerAttemptMetrics{{Fingerprint: fp}}},
		{Sequence: 3, PredecessorSequence: 2, ThreadID: "thread", Status: "completed", ProviderAttempts: []providerAttemptMetrics{{Fingerprint: fp}}},
	}
	diagnoseCacheExchanges(exchanges)
	if got := exchanges[1].CacheDiagnosis; got.PreviousSequence != 0 || got.Provider.Status != "unavailable" {
		t.Fatalf("skipped pending predecessor: %+v", got)
	}
}

func TestCacheDiagnosisDoesNotBridgeEvictedFailedPredecessor(t *testing.T) {
	r := diagnosticRecorder(t)
	header := http.Header{"Thread-Id": []string{"thread"}}
	first, _ := r.beginRequest(header)
	failed, _ := r.beginRequest(header)
	third, _ := r.beginRequest(header)
	if failed.predecessorSequence != first.sequence || third.predecessorSequence != failed.sequence {
		t.Fatal("arrival predecessor was not retained")
	}
	fp := r.requestFingerprint([]byte(`{"input":["stable"]}`))
	// The failed middle completion has left the bounded detail window before
	// the first and third requests complete. Its identity must still block reuse.
	exchanges := []exchangeMetrics{
		{Sequence: third.sequence, PredecessorSequence: third.predecessorSequence, ThreadID: "thread", Status: "completed", ProviderAttempts: []providerAttemptMetrics{{Fingerprint: fp}}},
		{Sequence: first.sequence, ThreadID: "thread", Status: "completed", ProviderAttempts: []providerAttemptMetrics{{Fingerprint: fp}}},
	}
	diagnoseCacheExchanges(exchanges)
	if got := exchanges[0].CacheDiagnosis; got.PreviousSequence != 0 || got.Provider.Status != "unavailable" {
		t.Fatalf("bridged evicted failed predecessor: %+v", got)
	}
}

func TestTurnStateFingerprintForwarding(t *testing.T) {
	r := diagnosticRecorder(t)
	for _, test := range []struct{ client, provider, want string }{
		{"", "", "absent"}, {"private-state", "private-state", "preserved"},
		{"private-state", "", "dropped"}, {"private-state", "different-state", "changed"},
		{"", "injected-state", "changed"},
	} {
		t.Run(test.want+test.provider, func(t *testing.T) {
			transport := r.Transport(roundTripperFunc(func(request *http.Request) (*http.Response, error) {
				if request.Header.Get("x-codex-turn-state") != test.provider {
					t.Error("capture changed forwarded header")
				}
				return &http.Response{StatusCode: 200, Header: http.Header{"Content-Type": []string{"application/json"}}, Body: io.NopCloser(strings.NewReader(`{"status":"completed","output":[]}`))}, nil
			}))
			handler := r.Handler(http.HandlerFunc(func(w http.ResponseWriter, incoming *http.Request) {
				body, _ := io.ReadAll(incoming.Body)
				req, _ := http.NewRequestWithContext(incoming.Context(), http.MethodPost, "http://provider/responses", bytes.NewReader(body))
				if test.provider != "" {
					req.Header.Set("x-codex-turn-state", test.provider)
				}
				response, err := transport.RoundTrip(req)
				if err != nil {
					t.Error(err)
					return
				}
				defer response.Body.Close()
				w.Header().Set("Content-Type", "application/json")
				_, _ = io.Copy(w, response.Body)
			}))
			incoming := httptest.NewRequest(http.MethodPost, "/v1/responses", strings.NewReader(`{"input":[]}`))
			// A current-request comparison must also work without thread identity.
			if test.want != "dropped" {
				incoming.Header.Set("thread-id", "thread")
			}
			if test.client != "" {
				incoming.Header.Set("x-codex-turn-state", test.client)
			}
			handler.ServeHTTP(httptest.NewRecorder(), incoming)
			snapshot := r.snapshot()
			latest := snapshot.Exchanges[len(snapshot.Exchanges)-1]
			if latest.CacheDiagnosis.TurnStateForwarding != test.want {
				t.Fatalf("forwarding status: %s, want %s", latest.CacheDiagnosis.TurnStateForwarding, test.want)
			}
			encoded, _ := json.Marshal(snapshot)
			for _, secret := range []string{"private-state", "different-state", "injected-state"} {
				if bytes.Contains(encoded, []byte(secret)) {
					t.Fatal("raw turn state retained")
				}
			}
			clone := cloneFingerprint(latest.ClientFingerprint)
			*clone.TurnState = "mutated"
			if *latest.ClientFingerprint.TurnState == "mutated" {
				t.Fatal("aliased turn-state fingerprint")
			}
		})
	}
	if compareTurnState(nil, nil) != "unavailable" {
		t.Fatal("missing evidence is not unavailable")
	}
	first := r.requestFingerprint([]byte(`{}`))
	if compareTurnState(first, first) != "unavailable" {
		t.Fatal("legacy evidence is not unavailable")
	}
}

func TestCacheFingerprintIncrementalHistoryIsNotAChangedPrefix(t *testing.T) {
	r := diagnosticRecorder(t)
	full := r.requestFingerprint([]byte(`{"model":"model","input":["base"]}`))
	first := r.requestFingerprint([]byte(`{"type":"response.create","model":"model","previous_response_id":"first-private-id","input":["suffix"]}`))
	next := r.requestFingerprint([]byte(`{"type":"response.create","model":"model","previous_response_id":"second-private-id","input":["different suffix"]}`))
	for _, pair := range [][2]*requestFingerprint{{full, first}, {first, next}, {next, full}} {
		got := comparePrefix(pair[0], pair[1])
		if got.Status != "unavailable" || got.CommonItems != 0 || len(got.ChangedFields) != 0 {
			t.Fatalf("wire suffix claimed a changed prefix: %+v", got)
		}
	}
	if !cloneFingerprint(next).Incremental {
		t.Fatal("clone lost incremental marker")
	}
	changed := r.requestFingerprint([]byte(`{"model":"different","tools":[{"name":"new"}],"previous_response_id":"third","input":[]}`))
	got := comparePrefix(next, changed)
	if got.Status != "unavailable" || len(got.ChangedFields) != 2 || got.ChangedFields[0] != "model" || got.ChangedFields[1] != "tools" {
		t.Fatalf("actual setting changes were hidden: %+v", got)
	}
	warm := r.requestFingerprint([]byte(`{"model":"model","generate":false,"input":["base"]}`))
	if got := comparePrefix(warm, full); got.Status != "identical" {
		t.Fatalf("prewarm transport flag changed inference fingerprint: %+v", got)
	}
	encoded, err := json.Marshal(next)
	if err != nil || strings.Contains(string(encoded), "private-id") {
		t.Fatalf("continuation identity leaked: %s, %v", encoded, err)
	}
}

func TestCacheAttributionLabelsItsEstimate(t *testing.T) {
	snapshot := newMetricsSnapshot("mekugi")
	if snapshot.Cache.AttributionBasis != "previous_input_length_estimate" {
		t.Fatal("cache attribution is not explicitly labeled as an estimate")
	}
}

File: benchmarks/cache_diagnostics_test.py
#!/usr/bin/env python3
"""Model-free checks for sanitized cache-diagnostic evidence."""
import copy
import sys
import unittest

sys.dont_write_bytecode = True
from analyze_capture import compare_prefix, validate_cache_diagnostics, validate_fingerprint, validate_provider_evidence


def fingerprint(items):
    return {"scope": "a" * 32, "fields": {"model": "b" * 32, "other":"f" * 32},
            "input_kind": "array", "items": items, "item_count": len(items),
            "complete": True, "routing_key": "c" * 32, "request_key": "c" * 32}


def exchange(sequence, fp, diagnosis):
    return {"sequence": sequence, "predecessor_sequence": sequence-1, "thread_id": "thread", "status": "completed",
            "client_fingerprint": fp, "provider_attempts": [
                {"cache_fingerprint": fp, "projected_fingerprint": fp}],
            "cache_diagnostics": diagnosis}


class CacheDiagnosticsTests(unittest.TestCase):
    def test_ordered_stage_comparison_and_tampering(self):
        a, b = fingerprint(["d" * 32]), fingerprint(["d" * 32, "e" * 32])
        unavailable = {"status": "unavailable", "common_items": 0, "changed_fields": []}
        appended = {"status": "appended", "common_items": 1, "changed_fields": []}
        first = {"previous_sequence": 0, "client": unavailable, "projected": unavailable,
                 "provider": unavailable, "routing": "unavailable", "request_key": "unavailable"}
        second = {"previous_sequence": 1, "client": appended, "projected": appended,
                  "provider": appended, "routing": "stable", "request_key": "stable"}
        rows = [exchange(2, b, second), exchange(1, a, first)]
        validate_cache_diagnostics(rows)
        broken = copy.deepcopy(rows)
        broken[0]["cache_diagnostics"]["routing"] = "changed"
        with self.assertRaises(ValueError):
            validate_cache_diagnostics(broken)
        broken = copy.deepcopy(rows)
        broken[0]["provider_attempts"][0]["cache_fingerprint"]["items"][0] = "f" * 32
        with self.assertRaises(ValueError):
            validate_cache_diagnostics(broken)

    def test_turn_state_forwarding_and_tampering(self):
        unavailable = {"status": "unavailable", "common_items": 0, "changed_fields": []}
        for client, provider, expected in (("", "", "absent"), ("d"*32, "d"*32, "preserved"),
                                           ("d"*32, "", "dropped"), ("d"*32, "e"*32, "changed")):
            fp = fingerprint([])
            fp["turn_state"] = client
            diagnosis = {"previous_sequence": 0, "client": unavailable, "projected": unavailable,
                         "provider": unavailable, "routing": "unavailable", "request_key": "unavailable",
                         "turn_state_forwarding": expected}
            row = exchange(1, fp, diagnosis)
            row["provider_attempts"][0]["cache_fingerprint"] = copy.deepcopy(fp)
            row["provider_attempts"][0]["cache_fingerprint"]["turn_state"] = provider
            validate_cache_diagnostics([row])
            del row["thread_id"]
            validate_cache_diagnostics([row])
            row["cache_diagnostics"]["turn_state_forwarding"] = "wrong"
            with self.assertRaises(ValueError):
                validate_cache_diagnostics([row])
        for invalid in (None, "private state", 42):
            fp = fingerprint([])
            fp["turn_state"] = invalid
            with self.assertRaises(ValueError):
                validate_fingerprint(fp)

    def test_provider_response_evidence(self):
        for state in ("missing", "null", "invalid", "unavailable"):
            e = {"cached_tokens_state": state}
            validate_provider_evidence(e, None)
            e["cached_tokens"] = 0
            with self.assertRaises(ValueError):
                validate_provider_evidence(e, None)
        for count in (0, 128):
            e = {"cached_tokens_state": "present", "cached_tokens": count,
                 "request_id": "req-test", "model": "response-model", "header_model": "header-model"}
            validate_provider_evidence(e, {"cached_input_tokens": count})
            with self.assertRaises(ValueError):
                validate_provider_evidence(e, {"cached_input_tokens": count+1})
        for changes in ({"cached_tokens": None}, {"cached_tokens": -1}, {"request_id": "unsafe\ntext"},
                        {"authorization": "secret"}, {"model": "x"*257}):
            e = {"cached_tokens_state": "present", "cached_tokens": 0} | changes
            with self.assertRaises(ValueError):
                validate_provider_evidence(e, None)
        validate_provider_evidence(None, {"cached_input_tokens": 0})

    def test_private_shape_and_partial_evidence(self):
        fp = fingerprint(["d" * 32])
        validate_fingerprint(fp)
        partial = copy.deepcopy(fp)
        partial.update(item_count=129, complete=False, items=["d" * 32]*128)
        self.assertEqual(compare_prefix(fp, partial)["status"], "unavailable")
        malformed = copy.deepcopy(fp)
        malformed["input_kind"] = "absent"
        with self.assertRaises(ValueError):
            validate_fingerprint(malformed)
        malformed = copy.deepcopy(fp)
        del malformed["fields"]["other"]
        with self.assertRaises(ValueError):
            validate_fingerprint(malformed)
        unsafe = copy.deepcopy(fp)
        unsafe["routing_key"] = "raw secret key"
        with self.assertRaises(ValueError):
            validate_fingerprint(unsafe)
        unsafe = copy.deepcopy(fp)
        unsafe["fields"]["private arbitrary key"] = "d" * 32
        with self.assertRaises(ValueError):
            validate_fingerprint(unsafe)


if __name__ == "__main__":
    unittest.main()
